### PR TITLE
libreoffice: Convert generate-libreoffice-srcs from Bash to Python

### PR DESCRIPTION
**Pull request by chris-martin**
This is a rewrite of `generate-libreoffice-srcs.sh` in Python. The output of this script is identical to that of its predecessor.

I think the result is easier to follow, and slightly less fragile. It does the variable substitution in a more general way, and relies less on the file's ordering.

It still prints skipped lines (and also line numbers) to stderr to help detect problems, though it doesn't print skipped comments.

### CODE REVIEW

File: pkgs/applications/office/libreoffice/default-gen-shell.nix [hunk 1]
```diff
@@ -19,9 +19,11 @@ stdenv.mkDerivation {
     builder = ./download-list-builder.sh;
   };
 
+  buildInputs = [ python36 ];
+
   shellHook = ''
     function generate {
-      ./generate-libreoffice-srcs.sh | tee libreoffice-srcs.nix
+      python3 generate-libreoffice-srcs.py > libreoffice-srcs.nix
```
Comment (7c6f434c) [inline pkgs/applications/office/libreoffice/default-gen-shell.nix]: Why not add `#! /usr/bin/env python3` in the script, too? I think most NixPkgs developers have `/usr/bin/env` (it is default-on in NixOS)

File: pkgs/applications/office/libreoffice/generate-libreoffice-srcs.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+    A mapping from package name (the all-caps identifiers used in
+    `download.lst`) to a dict of additional attributes to set on the package.
+    """
+    with open('./libreoffice-srcs-additions.json') as f:
+        return json.load(f)
+
+
+def apply_additions(xs, additions):
+    for x in xs:
+        yield {**x, **additions.get(x['name'], {})}
```
Comment (7c6f434c) [inline pkgs/applications/office/libreoffice/generate-libreoffice-srcs.py]: I don't know, maybe if the problem you want to fix is low readability of the bash script to people with weak knowledge of bash, `**x` thing in Python could be worth a comment like «merge two hashtables into one» for people who don't write Python.

File: pkgs/applications/office/libreoffice/default-gen-shell.nix [hunk 3]
```diff
@@ -19,9 +19,11 @@ stdenv.mkDerivation {
     builder = ./download-list-builder.sh;
   };
 
+  buildInputs = [ python36 ];
```
Comment (FRidh) [inline pkgs/applications/office/libreoffice/default-gen-shell.nix]: Why use a pre-release of Python 3.6? Please use `python3` which at this time refers to `python35`.

Comment (chris-martin) [inline pkgs/applications/office/libreoffice/default-gen-shell.nix]: Ah, I didn't know 3.6 was considered pre-release. I think I just wanted string interpolation. Guess we can do without it.

File: pkgs/applications/office/libreoffice/default-gen-shell.nix (hunk 1 above)
Comment (chris-martin) [inline pkgs/applications/office/libreoffice/default-gen-shell.nix]: I thought the lesson learned from NixOS was that shebangs were somewhat evil and `/usr/bin/env` was just a least-painful compromise. No problem adding it if you like, though.

File: pkgs/applications/office/libreoffice/generate-libreoffice-srcs.py (hunk 2 above)
Comment (chris-martin) [inline pkgs/applications/office/libreoffice/generate-libreoffice-srcs.py]: Fair, will do.

**Comment by chris-martin** [comment]
Added a shebang, changed package `python36` -> `python3`, removed all usages of the new `**` dict splat operator.

**Review by 7c6f434c — APPROVED** [review]
Fine with me. Any comments, @FRidh ?

**Review by FRidh — APPROVED** [review]

File: pkgs/applications/office/libreoffice/default-gen-shell.nix (hunk 1 above)
Comment (the-kenny) [inline pkgs/applications/office/libreoffice/default-gen-shell.nix]: I agree - there's no harm in adding a shebang. I'd still leave the explicit call to `python3` in this line though.

Comment (7c6f434c) [inline pkgs/applications/office/libreoffice/default-gen-shell.nix]: This is what happens in the current version.

**Pull request merged**